fxa-rust-client/src/lib.rs

```diff
@@ -0,0 +1,502 @@
+        }
+        let oauth_flow = match self.flow_store.remove(state) {
+            Some(oauth_flow) => oauth_flow,
+            None => bail!(ErrorKind::UnknownOAuthState),
+        };
+        self.handle_oauth_token_response(resp, oauth_flow.jwk)
+    }
+
+    fn handle_oauth_token_response(
+        &mut self,
+        resp: OAuthTokenResponse,
+        jwk: Option<JWK>,
+    ) -> Result<OAuthInfo> {
+        let granted_scopes = resp.scope.split(" ").map(|s| s.to_string()).collect();
+        // This assumes that if the server returns keys_jwe, the jwk argument is Some.
+        let keys_jwe = match resp.keys_jwe {
+            Some(jwe) => {
+                let jwk = jwk.expect("Insane state!");
+                let jwe = JWE::import(&jwe)?;
+                Some(jwk.decrypt(&jwe)?)
+            }
+            None => None,
+        };
+        let since_epoch = SystemTime::now()
+            .duration_since(UNIX_EPOCH)
+            .expect("Something is very wrong.");
+        let expires_at = since_epoch.as_secs() + resp.expires_in;
+        let oauth_info = OAuthInfo {
+            access_token: resp.access_token,
+            keys_jwe,
+            refresh_token: resp.refresh_token,
+            expires_at,
+            scopes: granted_scopes,
+        };
+        self.oauth_cache_store(&oauth_info);
+        Ok(oauth_info)
+    }
+
+    fn random_base64_url_string(len: usize) -> String {
+        let mut r = OsRng::new().expect("Could not instantiate RNG");
+        let mut buf: Vec<u8> = vec![0; len];
+        r.fill_bytes(buf.as_mut_slice());
+        let random = base64::encode_config(&buf, base64::URL_SAFE_NO_PAD);
+        random
+    }
+
+    fn session_token_from_state(state: &LoginState) -> Option<&[u8]> {
+        match state {
+            &Separated(_) | Unknown => None,
+            // Despite all these states implementing the same trait we can't treat
+            // them in a single arm, so this will do for now :/
+            &EngagedBeforeVerified(ref state) | &EngagedAfterVerified(ref state) => {
+                Some(state.session_token())
+            }
+            &CohabitingBeforeKeyPair(ref state) => Some(state.session_token()),
+            &CohabitingAfterKeyPair(ref state) => Some(state.session_token()),
+            &Married(ref state) => Some(state.session_token()),
+        }
+    }
+
+    pub fn generate_assertion(&mut self, audience: &str) -> Result<String> {
+        let married = match self.to_married() {
+            Some(married) => married,
+            None => bail!(ErrorKind::NotMarried),
+        };
+        let key_pair = married.key_pair();
+        let certificate = married.certificate();
+        Ok(jwt_utils::create_assertion(
+            key_pair,
+            &certificate,
+            audience,
+        )?)
+    }
+
+    pub fn get_profile(
+        &mut self,
+        profile_access_token: &str,
```

Comment:
As mentioned above, ISTM that since the `FirefoxAccount` owns the oauth tokens, I shouldn't have to pass one in explicitly here.

Reply:
Especially if the caller has to go to the trouble of creating a token to pass in to this function, but then it turns out the profile info was in cache and their effort was wasted.